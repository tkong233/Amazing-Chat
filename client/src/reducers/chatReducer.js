import { CONNECT_SOCKET, JOIN_ROOM, LOAD_PAST_MESSAGES,
   DISCONNECT_SOCKET, SEND_MESSAGE, RECEIVE_NEW_MESSAGES } from '../actions/types';

const initialState = {
  socket: null,
  pairId: null,
  sender: null,
  receiver: null,
  messages: [], // [{ pairId, from, to, message, datetime }]
}

export default function(state = initialState, action) {
  switch (action.type) {
    case CONNECT_SOCKET:
      return {
        ...state,
        socket: action.payload
      };
    case JOIN_ROOM:
      return {
        ...state,
        pairId: action.payload.pairId,
        sender: action.payload.sender,
        receiver: action.payload.receiver,
      };
    case LOAD_PAST_MESSAGES:
      let newMessages = state.messages;
      newMessages = action.payload;
      return {
        ...state,
        messages: newMessages,
      };
    case SEND_MESSAGE:
      newMessages = state.messages;
      newMessages.push(action.payload);
      console.log('send message', newMessages);
      return {
        ...state,
        messages: newMessages,
      };
    case RECEIVE_NEW_MESSAGES:
      console.log('receive new message reducer,', action.payload);
      if (action.payload.from !== state.sender) {
        newMessages = state.messages;
        newMessages.push(action.payload);
        console.log('added new message', newMessages);
        return {
          ...state,
          messages: newMessages,
        }
      } else {
        console.log('message was sent by myself!', state.messages);
        return {
          ...state,
        }
      }
    case DISCONNECT_SOCKET:
      return {
        ...state,
        socket: null,
      }
    default:
      return state;
  }
}