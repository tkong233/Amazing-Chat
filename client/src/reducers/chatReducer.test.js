import { SET_SOCKET, UNSET_SOCKET, CONNECT_SOCKET, JOIN_ROOM, LOAD_PAST_MESSAGES,
    DISCONNECT_SOCKET, SEND_MESSAGE, DELETE_MESSAGE, RECEIVE_NEW_MESSAGES,
    RECEIVE_VIDEO_CALL, PICK_UP_VIDEO_CALL, HANG_UP_VIDEO_CALL,
    INITIATE_VIDEO_CALL, SET_CALLEE_ONLINE, SET_CALLEE_OFFLINE,
    STOP_WAITING_FOR_CALLEE_RESPONSE, SET_ITEM_STATUS, VIDEO_CALL_REJECTED, RESET_VIDEO_CALL_REJECTED, 
    CALLER_HANGED_CALL, DELETE_CONVERSATION
   } from '../actions/types';
import chatReducer from './chatReducer';

describe('Auth Reducer', ()=>{
    it('Should return default state', ()=>{
        const newState = chatReducer(undefined, {});
        expect(newState).toEqual({
            socket: null,
            pairId: null,
            sender: null, //email
            receiver: null, // email
            senderName: null,
            receiverName: null,
            messages: [], // [{ pairId, from, to, message, datetime, type }]
            itemStatusMap: {}, // {index: status}, e.g. { 1 : LOADING, 2 : LOADED... }
            ringing: false,
            calling: false,
            waiting: false,
            rejected: false,
            calleeOnline: true,
        });
    });
    it ('join room', () =>{
        // pairId, sender, receiver, senderName, receiverName
        const chat = {
            socket: null,
            pairId: 'testid',
            sender: 'testsender', //email
            receiver: 'test receiver', // email
            senderName: 'test sender name',
            receiverName: 'test reciever name',
            // messages: [
            //     {pairId: 'testid', from: 'testid', to: 'testsender', message: 'hey', datetime:'2020/09/01', type:'text'} 
            // ], 
            messages: [],
            itemStatusMap: {}, // {index: status}, e.g. { 1 : LOADING, 2 : LOADED... }
            ringing: false,
            calling: false,
            waiting: false,
            rejected: false,
            calleeOnline: true,
        };
        const newState = chatReducer(undefined, {
            type: JOIN_ROOM,
            payload: chat
        });
        expect(newState).toEqual(chat);
    });

    it ('SET_ITEM_STATUS', () =>{
        const chat = {
            socket: null,
            pairId: null,
            sender: null, //email
            receiver: null, // email
            senderName: null,
            receiverName: null,
            messages: [], 
            itemStatusMap: {}, // {index: status}, e.g. { 1 : LOADING, 2 : LOADED... }
            ringing: false,
            calling: false,
            waiting: false,
            rejected: false,
            calleeOnline: true,
        };
        const newState = chatReducer(undefined, {
            type: SET_ITEM_STATUS,
            payload: chat
        });
        expect(newState).toEqual(chat);
    });

    // it ('LOAD_PAST_MESSAGES', () =>{
    //     const chat1 = {
    //         socket: null,
    //         pairId: null,
    //         sender: null, //email
    //         receiver: null, // email
    //         senderName: null,
    //         receiverName: null,
    //         messages: [
    //             {pairId: 'testid', from: 'testid', to: 'testsender', message: 'hey', datetime:'2020/09/01', type:'text'} 
    //         ], 
    //         itemStatusMap: {}, // {index: status}, e.g. { 1 : LOADING, 2 : LOADED... }
    //         ringing: false,
    //         calling: false,
    //         waiting: false,
    //         rejected: false,
    //         calleeOnline: true,
    //     }
    //     const chat = {
    //         socket: null,
    //         pairId: null,
    //         sender: null, //email
    //         receiver: null, // email
    //         senderName: null,
    //         receiverName: null,
    //         messages: [
    //             {pairId: 'testid', from: 'testid', to: 'testsender', message: 'hey', datetime:'2020/09/01', type:'text'},
    //             {pairId: 'testid', from: 'testid', to: 'testsender', message: 'heyhey', datetime:'2020/09/01', type:'text'}  
    //         ], 
    //         itemStatusMap: {}, // {index: status}, e.g. { 1 : LOADING, 2 : LOADED... }
    //         ringing: false,
    //         calling: false,
    //         waiting: false,
    //         rejected: false,
    //         calleeOnline: true,
    //     };
    //     const newState = chatReducer(chat, {
    //         type: LOAD_PAST_MESSAGES,
    //         payload: chat
    //     });
    //     expect(newState).toEqual(chat);
    // });
    it ('should return initial state', ()=>{
        expect(chatReducer(undefined, {})).toMatchSnapshot();
    });
});