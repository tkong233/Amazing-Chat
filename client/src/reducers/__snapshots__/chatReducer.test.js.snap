// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Reducer should return initial state 1`] = `
Object {
  "calleeOnline": true,
  "calling": false,
  "itemStatusMap": Object {
    "undefined": undefined,
  },
  "messages": Array [],
  "pairId": null,
  "receiver": null,
  "receiverName": null,
  "rejected": false,
  "ringing": false,
  "sender": null,
  "senderName": null,
  "socket": null,
  "waiting": false,
}
`;
